### PR TITLE
[Feature/Identity] Allow for Encryption/Decryption of Principals into PITs 

### PR DESCRIPTION
**Pull request by scrawfor99**
Changes in the ExtensionTokenProcessor class to allow for encryption & decryption of principal & extension identifiers. 

Tests not yet added but will add unit tests. Right now there are some lingering issues with the code that I would appreciate feedback or thoughts on:

As you can see at Line 146, the ExtractPrincipal method attempts to return the Principal object that is extracted from the encrypted message. Right now, I am not sure that we are actually storing the Principals themselves in any way so to be able to return an entire Principal instance we would need to be able to store and then retrieve the instance via NAME lookup.

I am also not entirely convinced by my implementation of the encoding combining at Line 106. I know you can combine byte[]s in this manner but I have not had a chance to confirm this preserves the encoding integrity as expected.

Resolves 
- Issue #4485 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
@scrawfor99 Can you prepend the title of this PR with `[Feature/Identity]`? It makes it easier to find PRs against this feature branch and core maintainers can see from the description that its for the feature branch.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:124]: I know this is a typo, but I think you just coined a new term ;)

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:49]: The `extensionUniqueId` should remain a String.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:132]: Variable names should be in `lowerCamelCase`

File: server/src/main/java/org/opensearch/identity/Principals.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/Principals.java:44]: Are these methods duplicates?

**Comment by scrawfor99** [comment]
> @scrawfor99 Can you prepend the title of this PR with `[Feature/Identity]`? It makes it easier to find PRs against this feature branch and core maintainers can see from the description that its for the feature branch.

Should be all set

File: server/src/main/java/org/opensearch/identity/Principals.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/Principals.java:35]: Won't this generate keys for every instance of this class? Even if this is a singleton class, wouldn't this generate a different key for every node of the cluster? 

We may need to look into how [`SecureSetting`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/settings/SecureSetting.java) stores secrets and think about storage and configuration for these keys.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:128]: Can you write tests for this to shows its usage?

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:96]: Can you write tests for this to show its usage?

**Review by cwperks — CHANGES_REQUESTED** [review]
Thank you for the PR @scrawfor99! If you want to submit a PR to solicit initial feedback you can use Github's `Draft PR` feature to distinguish the PR as a WIP (work in progress).   

Can you add tests for the functionality added in this PR?

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:49]: It has to be a byte[] to encode it but I switched the name of the attribute to avoid the confusion with the argument being passed in (which will be a string as you mentioned).

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:132]: Oops, old Python habits diehard.

File: server/src/main/java/org/opensearch/identity/Principals.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/Principals.java:35]: It will generate a new key and initialization vector for every principal. This is the only way I am aware of to make sure that each principal is provided a different encoding. Otherwise, the encodings would only be different for each extension but would remain the same for all the different principals on a given extension. 

I take it this is not what we want? 

For the creation of the instances on each node, I am not sure how to handle this as currently I do not know what type of data sharing between nodes we have--I can look at the SecureSettings class but we do not currently have any principal lookup or storage system that I am aware of.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:128]: yeah, I was just hoping to put this out there as soon as possible because I know we are trying to push rapidly & everyone's feedback is invaluable. I do plan to add tests of course and appreciate all the suggestions.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:96]: Definitely!

File: server/src/main/java/org/opensearch/identity/Principals.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/identity/Principals.java:35]: Ignore this. I am not sure why the PR continues to show this file change when it no longer exists in the merging branch but these changes are gone. Sorry for the confusion.

**Review by DarshitChanpura — CHANGES_REQUESTED** [review]
Ty @scrawfor99 for your research and this PR!  I have requested few changes around how the tests are structured/modified.

File: buildSrc/src/integTest/resources/org/opensearch/gradle/internal/fake_git/remote/settings.gradle
Comment (peternied) [inline buildSrc/src/integTest/resources/org/opensearch/gradle/internal/fake_git/remote/settings.gradle:31]: Can we revert this change?

**Review by peternied — CHANGES_REQUESTED** [review]
Can we rollback the gradle change? While I've got outstanding feedback we can address that in a future PR

**Comment by scrawfor99** [comment]
> Can we rollback the gradle change? While I've got outstanding feedback we can address that in a future PR

Done!

**Comment by peternied** [comment]
Looks like there are still some build issues, can you resolve those before we merge?

**Comment by scrawfor99** [comment]
> Looks like there are still some build issues, can you resolve those before we merge?

I have been looking at this but I am not sure how to fix the gradle issue. It is the result of missing dependencies in qa/full-cluster-restart -- but I am not sure what the appropriate way to introduce these dependencies or if I am missing something.

**Comment by peternied** [comment]
@scrawfor99 Follow that Gradle Check failure link [1], click 'Console Output', then click 'Full Log" and look for an error, it shows the console log from the `./gradle check` command.  I pulled out what I believe are the only errors.  You should be able to rerun locally or just rerun the task that failed to confirm you've resolved it.  

```
> Task :server:forbiddenApisMain
Forbidden method invocation: java.lang.String#getBytes() [Uses default charset]
  in org.opensearch.identity.ExtensionTokenProcessor (ExtensionTokenProcessor.java:103)
Forbidden method invocation: java.lang.String#getBytes() [Uses default charset]
  in org.opensearch.identity.ExtensionTokenProcessor (ExtensionTokenProcessor.java:105)
Forbidden method invocation: java.lang.String#format(java.lang.String,java.lang.Object[]) [Uses default locale]
  in org.opensearch.identity.ExtensionTokenProcessor (ExtensionTokenProcessor.java:133)
```


[1] https://build.ci.opensearch.org/job/gradle-check/5111/

**Comment by scrawfor99** [comment]
> 

Thank you for helping find this. I had no idea there was this external view of the tests! I will look at these issues and get them fixed. It seems like it just does not want me using one library for whatever reason so I will check it out.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (DarshitChanpura) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:182]: this is causing forbiddenApis to fail. All instances where byte[] is converted to String will throw this

**Comment by peternied** [comment]
> Everything seems fine on my end but now the check is failing because of memory errors from other code. 

Sounds like there is as bug in OpenSearch, can you capture the error output and create a bug like the following?
- https://github.com/opensearch-project/OpenSearch/issues/4852

**Comment by scrawfor99** [comment]
Assuming the Check passes (it should), this is ready for merge @cwperks @peternied @DarshitChanpura

**Comment by peternied** [comment]
@scrawfor99 I merged this because the failed tests are unrelated to your changes.  Next merge from main should resolve them (also what a pain!)  Thanks for your contribution and stick-with-it-ness on this one!

**Pull request merged**
